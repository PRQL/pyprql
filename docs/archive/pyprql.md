# Source Code

Currently, the source code is divided as follows:

- `magic` module that implements the PRQL IPython/Jupyer magic.
  See [IPython and Jupyter Magic](./magic.md).

```{toctree}
:hidden:
:maxdepth: 3

magic
```
