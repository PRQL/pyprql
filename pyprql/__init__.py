# -*- coding: utf-8 -*-
"""Source code for PyPrql."""
import prql_python as prql  # noqa: F401
from prql_python import to_sql as to_sql  # noqa: F401

__version__ = "0.5.0"
