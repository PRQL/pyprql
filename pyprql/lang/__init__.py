# -*- coding: utf-8 -*-
"""The language specification for PyPRQL."""
