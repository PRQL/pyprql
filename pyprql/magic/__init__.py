# -*- coding: utf-8 -*-
"""IPython/Jupyter magic for PyPRQL.

Examples
--------
A single function is defined herein.
It should **_never_** be used directly by the user.
Rather,
any users should load the magic using the IPython line magic,
like below:

    In [1]: %load_ext pyprql.magic

"""
from IPython import InteractiveShell

from .prql import PRQLMagic


def load_ipython_extension(ipython: InteractiveShell) -> None:
    """Load the ``pyprql.magic`` extension.

    This function is called automatically by ``IPython`` when
    the magic is loaded using ``%load_ext``.

    Parameters
    ----------
    ipython: InteractiveShell
        The current IPython instance.
    """
    ipython.register_magics(PRQLMagic)
