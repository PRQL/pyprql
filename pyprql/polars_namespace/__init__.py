from .prql import PrqlNamespace  # noqa: F401
