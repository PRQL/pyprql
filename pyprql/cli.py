# -*- coding: utf-8 -*-
import os
import sys
from typing import Dict, List, Optional

import pygments
import rich
from enforce_typing import enforce_types
from prompt_toolkit import prompt
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.completion import Completer, Completion
from prompt_toolkit.history import FileHistory
from prompt_toolkit.key_binding import KeyBindings
from prompt_toolkit.lexers import PygmentsLexer
from prompt_toolkit.styles import style_from_pygments_dict
from pygments.formatters.terminal import TerminalFormatter as Formatter
from pygments.lexers.sql import SqlLexer
from pygments.style import Style
from pygments.token import (
    Comment,
    Error,
    Generic,
    Keyword,
    Name,
    Number,
    Operator,
    String,
    Token,
    Whitespace,
)
from rich.table import Table
from sqlalchemy import create_engine, inspect

import pyprql.prql as prql
from pyprql.PRQLLexer import PRQLLexer

bindings = KeyBindings()
this_files_path = os.path.abspath(os.path.dirname(__file__))


@bindings.add("c-l")
def clear_screen(event):
    print(chr(27) + "[2j")
    print("\033c")
    print("\x1bc")


# Shortlist material ,pastie, trac, native, paraiso-dark, dracula

# PRQLStyle = style_from_pygments_cls(get_style_by_name('prql'))
class PRQLStyle(Style):
    """
    Pygments version of the "native" vim theme.
    """

    background_color = "#202020"
    highlight_color = "#404040"
    line_number_color = "#aaaaaa"

    styles = {
        Token: "#d0d0d0",
        Whitespace: "#666666",
        Comment: "italic #999999",
        Comment.Preproc: "noitalic bold #cd2828",
        Comment.Special: "noitalic bold #e50808 bg:#520000",
        Keyword: "bold #6ab825",
        Keyword.Pseudo: "nobold",
        Operator.Word: "bold #6ab825",
        String: "#ed9d13",
        String.Other: "#ffa500",
        Number: "#3677a9",
        Name.Builtin: "#24909d",
        Name.Variable: "#40ffff",
        Name.Constant: "#40ffff",
        Name.Class: "underline #447fcf",
        Name.Function: "#447fcf",
        Name.Namespace: "underline #447fcf",
        Name.Exception: "#bbbbbb",
        Name.Tag: "bold #6ab825",
        Name.Attribute: "#bbbbbb",
        Name.Decorator: "#ffa500",
        Generic.Heading: "bold #ffffff",
        Generic.Subheading: "underline #ffffff",
        Generic.Deleted: "#d22323",
        Generic.Inserted: "#589819",
        Generic.Error: "#d22323",
        Generic.Emph: "italic",
        Generic.Strong: "bold",
        Generic.Prompt: "#aaaaaa",
        Generic.Output: "#cccccc",
        Generic.Traceback: "#d22323",
        Error: "bg:#e3d2d2 #a61717",
    }


class PRQLCompleter(Completer):
    @enforce_types
    def __init__(
        self,
        table_names: List[str],
        column_names: List[str],
        column_map: Dict,
        prql_keywords: List[str],
    ):
        self.table_names = table_names
        self.column_names = column_names
        self.column_map = column_map
        self.prql_keywords = prql_keywords

        self.prev_word = None
        self.previous_selection = None

    def get_completions(self, document, complete_event):
        word_before_cursor = document.get_word_before_cursor(WORD=True)
        completion_operators = ["[", "+", ",", ":"]
        possible_matches = {
            "from": self.table_names,
            "join": self.table_names,
            "columns": self.table_names,
            "select": self.column_names,
            " ": self.column_names,
            "sort": self.column_names,
            "filter": self.column_names,
            "show": ["tables", "columns", "connection"],
            "exit": None,
        }
        for op in completion_operators:
            possible_matches[op] = self.column_names

        # This delays the completions until they hit space, or a completion operator
        if word_before_cursor in possible_matches:
            selection = possible_matches[word_before_cursor]
            selection = [f"{x}" for x in selection]
            self.previous_selection = selection
            if (
                word_before_cursor == "from"
                or word_before_cursor == "join"
                or word_before_cursor == "sort"
                or word_before_cursor == "select"
                or word_before_cursor == "columns"
                or word_before_cursor == "show"
                or word_before_cursor == ","
                or word_before_cursor == "["
                or word_before_cursor == "filter"
            ):
                pass
            else:
                for m in selection:
                    yield Completion(m, start_position=-len(word_before_cursor))

        # If its an operator
        elif (
            len(word_before_cursor) >= 1
            and word_before_cursor[-1] in completion_operators
        ):
            selection = possible_matches[word_before_cursor[-1]]
            self.previous_selection = selection
        # If its a period, then we assume the first word was a table
        elif len(word_before_cursor) >= 1 and word_before_cursor[-1] == ".":
            table = word_before_cursor[:-1]
            if table in self.column_map:
                selection = self.column_map[table]
                self.previous_selection = selection
                for m in selection:
                    yield Completion(m, start_position=0)
        # This goes back to the first if, this is the delayed completion finally completing
        elif self.previous_selection:
            selection = [
                x for x in self.previous_selection if x.find(word_before_cursor) != -1
            ]
            self.previous_selection = selection

            for m in selection:
                yield Completion(m, start_position=-len(word_before_cursor))

        # This is supposed to complete keywords, but its tromping all over the other completions
        # else:
        #     completer = FuzzyWordCompleter(self.prql_keywords)
        #     for m in completer.get_completions(document, complete_event):
        #         yield m


class CLI:
    def __init__(self, connect_str: str = "chinook"):
        self.has_one_blank = False
        self.prompt_text = "PRQL> "
        self.command = ""
        self.sql_mode = False
        self.connect_str = connect_str
        if connect_str == "chinook":
            connect_str = f"sqlite:///{this_files_path}/../resources/chinook.db"
        elif connect_str == "factbook":
            connect_str = f"sqlite:///{this_files_path}/../resources/factbook.db"

        rich.print(
            "Connecting to [pale_turquoise1]{}[/pale_turquoise1]".format(connect_str)
        )
        self.engine = create_engine(connect_str)
        self.inspector = inspect(self.engine)

    def get_all_columns(self):
        tables = self.inspector.get_table_names()
        columns = {}
        for table in tables:
            columns[table] = self.inspector.get_columns(table)
            columns[table] = [x["name"] for x in columns[table]]
            columns[table].sort()

        column_names = []
        for col in columns.keys():
            for column in columns[col]:
                column_names.append(column)
        column_names = list(set(column_names))
        column_names.sort()
        return column_names, columns

    def execute_sql(self, sql):
        with self.engine.connect() as con:
            rs = con.execute(sql)
            columns = rs.keys()
            table = Table(show_header=True, header_style="bold sandy_brown")
            for column in columns:
                table.add_column(column, justify="left")

            for _row in rs:
                row = list(_row)
                table.add_row(*[str(x) for x in row])

            rich.print(table)

    def highlight_prql(self, text):

        highlighted = pygments.highlight(text, PRQLLexer(), Formatter())
        return highlighted

    def highlight_sql(self, text):
        highlighted = pygments.highlight(text, SqlLexer(), Formatter())
        return highlighted

    @enforce_types
    def handle_input(self, _user_input: str) -> None:

        user_input: str = _user_input.strip().rstrip(";")
        if user_input == "prql":
            self.sql_mode = False
            self.prompt_text = "PRQL> "
            return
        elif user_input == "examples":
            rich.print(
                """
            [pale_turquoise1]SQL  : SELECT * from employees[/pale_turquoise1]
            [sandy_brown]PRQL : from employees[/sandy_brown]

            [pale_turquoise1]SQL  : SELECT name, salary from employees WHERE salary > 100000[/pale_turquoise1]
            [sandy_brown]PRQL : from employees | select \[name,salary] | filter salary > 100000[/sandy_brown]

            """
            )
            return
        elif user_input == "?" or user_input == "help":

            if self.sql_mode:
                rich.print(
                    "\tCommand [cornflower_blue bold]show tables[/cornflower_blue bold]: To show all tables in the database.\n"
                    + "\tCommand [cornflower_blue bold]show columns ${table}[/cornflower_blue bold]: To show all columns in a table.\n"
                )

            else:
                rich.print(
                    "\tCommand [cornflower_blue bold]show tables[/cornflower_blue bold]: To show all tables in the database.\n"
                    + "\tCommand [cornflower_blue bold]show columns ${table}[/cornflower_blue bold]: To show all columns in a table.\n"
                )

                rich.print(
                    "\tCommand [cornflower_blue bold]sql[/cornflower_blue bold]: Switch to SQL mode"
                )
                rich.print(
                    "\tCommand [cornflower_blue bold]prql[/cornflower_blue bold]: Switch to PRQL mode"
                )
                rich.print(
                    "\tCommand [cornflower_blue bold]<enter>[/cornflower_blue bold]: Hit enter twice to execute your query"
                )
                rich.print(
                    "\n\tCommand [cornflower_blue bold]examples[/cornflower_blue bold]: Displays sql-to-prql examples for reference"
                )

                self.prompt_text = "PRQL> "

            rich.print(
                "\nPRQL Syntax documentation is here https://github.com/max-sixty/prql\n"
            )

            return
        elif user_input == "sql":
            self.sql_mode = True
            self.prompt_text = "SQL> "
            return
        elif user_input == "show tables":
            # tables = self.engine.list_tables()
            tables = self.inspector.get_table_names()
            table = Table(show_header=True, header_style="bold sandy_brown")
            table.add_column("Table Name", justify="left")
            for table_name in tables:
                table.add_row(table_name)
            rich.print(table)
            return
        elif user_input.startswith("show columns"):
            table_name = user_input["show columns ".__len__() :]
            # tables = self.engine.list_tables()
            columns = self.inspector.get_columns(table_name)
            rich.print(columns)
            return

        self.command += user_input + " "
        if self.sql_mode:
            if not user_input:

                self.has_one_blank = False
                sql = self.command
                if "LIMIT" not in sql:
                    sql += " LIMIT 5"

                print("\t" + self.highlight_sql(sql))
                self.prompt_text = "SQL> "
                self.execute_sql(sql)

            else:
                self.prompt_text = "....>"
        else:
            if not user_input:
                self.has_one_blank = False
                if self.command and self.command.strip().rstrip("") != "":
                    sql = prql.to_sql(self.command)
                    if "LIMIT" not in sql:
                        sql += " LIMIT 5"

                    rich.print("[pale_green3 bold]SQL:[/pale_green3 bold]")
                    print("\t" + self.highlight_sql(sql))
                    rich.print(
                        "[medium_turquoise bold]Results:[/medium_turquoise bold]"
                    )
                    self.execute_sql(sql)
                    self.command = ""
                self.prompt_text = "PRQL> "

            else:
                self.prompt_text = "....>"

    def run(self):
        prql_keywords = [
            "select",
            "from",
            "filter",
            "derive",
            "aggregate",
            "sort",
            "take",
            "order",
        ]
        while True:
            all_columns, columns_map = self.get_all_columns()
            user_input = prompt(
                self.prompt_text,
                history=FileHistory(".prql-history.txt"),
                auto_suggest=AutoSuggestFromHistory(),
                completer=PRQLCompleter(
                    self.inspector.get_table_names(),
                    all_columns,
                    columns_map,
                    prql_keywords,
                ),
                lexer=PygmentsLexer(PRQLLexer),
                style=style_from_pygments_dict(PRQLStyle.styles),
            )
            try:
                self.handle_input(user_input)
            except Exception as e:
                print(
                    f"Exception when handling the input: {e},{repr(e)}\nContinuing..."
                )
                self.command = ""
                self.prompt_text = "PRQL> "
                self.has_one_blank = False


def print_usage():
    print(
        """
    Usage:
        python cli.py connection_string"""
    )

    print(
        """
    Examples:
        python cli.py 'sqlite:///file.db'
        python cli.py 'postgresql://user:password@localhost:5432/database'
        python cli.py 'postgresql+psycopg2://user:password@localhost:5432/database'
        python cli.py 'mysql://scott:tiger@localhost/foo'
        """
    )

    print(
        """
    Test Database:
        python cli.py chinook
        python cli.py factbook

    """
    )

    print(
        """
    Notes:
        The connection string syntax is detailed here https://docs.sqlalchemy.org/en/13/core/engines.html#database-urls
        To install database drivers, see https://docs.sqlalchemy.org/en/13/dialects/index.html

        Mysql      : pip install mysqlclient
        Postgresql : pip install psycopg2-binary
        MariaDB    : pip install mariadb
        Oracle     : pip install cx_oracle
        SQLite     : <built-in>
    """
    )


def main(params: Optional[List[str]] = None) -> None:
    if params is None:
        params = sys.argv
    try:
        if len(params) > 1:
            cli = CLI(params[1])
            cli.run()
        else:
            print_usage()
            sys.exit(0)
    except KeyboardInterrupt:
        sys.exit(0)
