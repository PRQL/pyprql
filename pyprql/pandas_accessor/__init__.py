"""Init file for pyprql.pandas.

Examples
--------
import  pandas as pd
import pyprql.pandas
df = pd.DataFrame({})
results_df = df.prql.query('select [age,name,occupation] | filter age > 21')

"""
import pyprql.pandas_accessor.prql  # noqa: F401
