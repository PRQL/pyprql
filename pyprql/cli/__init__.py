# -*- coding: utf-8 -*-
"""The command line interface implementation for PyPRQL."""
import sys
from typing import List, Optional

from pyprql.cli.cli import CLI


def main(params: Optional[List[str]] = None) -> None:
    """Serve the CLI entrypoint.

    If ``params`` is left as it's default ``None``,
    then ``params`` is set to ``sys.argv``.
    If no parameters are passed,
    then the help message is printed.
    Otherwise,
    a prompt is activated until a keyboard interrupt.

    Parameters
    ----------
    params : Optional[List[str]], default None
        The parameters passed to the CLI.
    """
    if params is None:
        params = sys.argv
    try:
        if len(params) > 1:
            cli = CLI(params[1])
            cli.run()
        else:
            CLI.print_usage()
            sys.exit(0)
    except KeyboardInterrupt:
        sys.exit(0)


if __name__ == "__main__":
    main()
