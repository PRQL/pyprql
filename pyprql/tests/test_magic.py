"""
These tests are heavily borrowed from
https://github.com/ploomber/jupysql/blob/0.5.1/src/tests/test_magic.py

Many are xfailed because we don't yet support them. Others are lightly adapted for PRQL,
and there are some TODOs around a fuller transition.

Thanks to @ploomber for providing the base for these tests as well as the extension.
"""


import os.path
import re
import tempfile
from textwrap import dedent

import polars as pl
import pytest
from sqlalchemy import create_engine

from pyprql.tests import run_sql

from . import run_prql


def test_memory_db(ip, regtest):
    print(run_prql(ip, "from test"), file=regtest)


def test_html(ip):
    result = run_prql(ip, "from test")
    assert "<td>foo</td>" in result._repr_html_().lower()


def test_line_magic(ip):
    result = ip.run_line_magic("prql", "from      test   |     select {name}")
    assert "foo" in str(result)


@pytest.mark.skip(reason="We only support pandas")
def test_print(ip):
    result = run_prql(ip, "from test")
    assert re.search(r"1\s+\|\s+foo", str(result))


@pytest.mark.skip(reason="We only support pandas")
def test_plain_style(ip):
    ip.run_line_magic("config", "SqlMagic.style = 'PLAIN_COLUMNS'")
    result = run_prql(ip, "from test")
    assert re.search(r"1\s+\|\s+foo", str(result))


@pytest.mark.skip
def test_multi_sql(ip):
    result = ip.run_cell_magic(
        "prql",
        "",
        """
        sqlite://
        from author
        """,
    )
    assert "Shakespeare" in str(result) and "Brecht" in str(result)


def test_result_var(ip, capsys):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")
    ip.run_cell_magic(
        "prql",
        "sqlite:// x <<",
        """
        from author
        select last_name
        """,
    )
    result = ip.user_global_ns["x"]
    out, _ = capsys.readouterr()

    assert "Shakespeare" in str(result) and "Brecht" in str(result)
    assert "Returning data to local variable" not in out


@pytest.mark.xfail(reason="Not supported in PRQL")
def test_result_var_multiline_shovel(ip):
    ip.run_cell_magic(
        "prql",
        "",
        """
        x << from author
        select last_name
        """,
    )
    result = ip.user_global_ns["x"]
    assert "Shakespeare" in str(result) and "Brecht" in str(result)


@pytest.mark.skip(reason="We only support pandas")
def test_access_results_by_keys(ip):
    assert run_prql(ip, "from author")["author"] == (
        "William",
        "Shakespeare",
        1616,
    )


def test_duplicate_column_names_accepted(ip):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")
    result = ip.run_cell_magic(
        "prql",
        "sqlite://",
        """
        from author | select {last_name, last_name}
        """,
    )
    assert ("Brecht", "Brecht") in result


def test_persist(ip):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")
    run_prql(ip, "")
    ip.run_cell("results = %prql from test")
    ip.run_cell("results_dframe = results.DataFrame()")
    ip.run_cell("%sql --persist sqlite:// results_dframe")
    persisted = run_prql(ip, "from results_dframe")
    assert persisted == [(0, 1, "foo"), (1, 2, "bar")]


def test_persist_no_index(ip):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")
    run_prql(ip, "")
    ip.run_cell("results = %prql from test")
    ip.run_cell("results_no_index = results.DataFrame()")
    ip.run_cell("%prql --persist sqlite:// results_no_index --no-index")
    persisted = run_prql(ip, "from results_no_index")
    assert persisted == [(1, "foo"), (2, "bar")]


def test_append(ip):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")
    run_prql(ip, "")
    ip.run_cell("results = %prql from test")
    ip.run_cell("results_dframe_append = results.DataFrame()")
    ip.run_cell("%prql --persist sqlite:// results_dframe_append")
    persisted = run_prql(ip, "from results_dframe_append")
    ip.run_cell("%prql --append sqlite:// results_dframe_append")
    appended = run_prql(ip, "from results_dframe_append")
    assert appended[0][0] == persisted[0][0] * 2


@pytest.mark.xfail(reason="Need to resolve line vs cell magic")
def test_persist_frame_at_its_creation(ip):
    ip.run_cell("results = %sql from author")
    ip.run_cell("%sql --persist sqlite:// results.DataFrame()")
    persisted = run_prql(ip, "from results")
    assert "Shakespeare" in str(persisted)


def test_connection_args_in_connection(ip):
    ip.run_cell('%sql --connection_arguments {"timeout":10} sqlite:///:memory:')
    result = ip.run_cell("%sql --connections")
    assert "timeout" in result.result["sqlite:///:memory:"].connect_args


def test_connection_args_single_quotes(ip):
    ip.run_cell("%sql --connection_arguments '{\"timeout\": 10}' sqlite:///:memory:")
    result = ip.run_cell("%sql --connections")
    assert "timeout" in result.result["sqlite:///:memory:"].connect_args


# TODO: support
# @with_setup(_setup_author, _teardown_author)
# def test_persist_with_connection_info():
#     ip.run_cell("results = %sql from author")
#     ip.run_line_magic('sql', 'sqlite:// PERSIST results.DataFrame()')
#     persisted = ip.run_line_magic('sql', 'SELECT * FROM results')
#     assert 'Shakespeare' in str(persisted)


@pytest.mark.xfail(reason="Need to implement sample data in PRQL")
def test_displaylimit(ip, regtest):
    ip.run_line_magic("config", "SqlMagic.autolimit = None")
    ip.run_line_magic("config", "SqlMagic.displaylimit = None")
    result = run_prql(
        ip,
        "SELECT * FROM (VALUES ('apple'), ('banana'), ('cherry')) "
        "AS Result ORDER BY 1;",
    )
    regtest.write(result._repr_html_(), file=regtest)

    assert "apple" in result._repr_html_()
    assert "banana" in result._repr_html_()
    assert "cherry" in result._repr_html_()

    ip.run_line_magic("config", "SqlMagic.displaylimit = 1")
    result = run_prql(
        ip,
        "SELECT * FROM (VALUES ('apple'), ('banana'), ('cherry')) "
        "AS Result ORDER BY 1;",
    )
    assert "apple" in result._repr_html_()
    assert "cherry" not in result._repr_html_()


@pytest.mark.xfail(reason="Not supported in PRQL")
def test_column_local_vars(ip):
    ip.run_line_magic("config", "SqlMagic.column_local_vars = True")
    result = run_prql(ip, "from author")
    assert result is None
    assert "William" in ip.user_global_ns["first_name"]
    assert "Shakespeare" in ip.user_global_ns["last_name"]
    assert len(ip.user_global_ns["first_name"]) == 2
    ip.run_line_magic("config", "SqlMagic.column_local_vars = False")


@pytest.mark.skip(reason="Not supported in PRQL")
def test_userns_not_changed(ip):
    ip.run_cell(
        dedent(
            """
    def function():
        local_var = 'local_val'
        %sql sqlite:// INSERT INTO test VALUES (2, 'bar');
    function()"""
        )
    )
    assert "local_var" not in ip.user_ns


@pytest.mark.xfail(reason="Not supported in PRQL")
def test_bind_vars(ip):
    ip.user_global_ns["x"] = 22
    result = run_prql(ip, "SELECT :x")
    assert result[0][0] == 22


def test_autopandas(ip):
    dframe = run_prql(ip, "from test")
    assert not dframe.empty
    assert dframe.ndim == 2
    assert dframe.name[0] == "foo"


def test_autopolars(ip):
    ip.run_line_magic("config", "PrqlMagic.autopolars = True")
    dframe = run_prql(ip, "from test")

    assert type(dframe) == pl.DataFrame
    assert not dframe.is_empty()
    assert len(dframe.shape) == 2
    assert dframe["name"][0] == "foo"


def test_target_dialect(ip):
    ip.run_line_magic("config", 'PrqlMagic.target = "sql.sqlite"')
    dframe = run_prql(
        ip, 'from author | select foo = f"{first_name}-{last_name}" | take 1'
    )
    assert dframe.foo[0] == "William-Shakespeare"


def test_dryrun(ip, capsys):
    ip.run_line_magic("config", "PrqlMagic.dryrun = True")
    result = run_prql(ip, 'from a | select b = f"{c}-{d}"')
    captured = capsys.readouterr()
    assert captured.out.startswith("SELECT\n  CONCAT")
    assert result is None


def test_csv(ip):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")  # uh-oh
    result = run_prql(ip, "from test")
    result = result.csv()
    for row in result.splitlines():
        assert row.count(",") == 1
    assert len(result.splitlines()) == 3


def test_csv_to_file(ip):
    ip.run_line_magic("config", "SqlMagic.autopandas = False")  # uh-oh
    result = run_prql(ip, "from test")
    with tempfile.TemporaryDirectory() as tempdir:
        fname = os.path.join(tempdir, "test.csv")
        output = result.csv(fname)
        assert os.path.exists(output.file_path)
        with open(output.file_path) as csvfile:
            content = csvfile.read()
            for row in content.splitlines():
                assert row.count(",") == 1
            assert len(content.splitlines()) == 3


def test_sql_from_file(ip):
    ip.run_line_magic("config", "PrqlMagic.autopandas = False")
    with tempfile.TemporaryDirectory() as tempdir:
        fname = os.path.join(tempdir, "test.sql")
        with open(fname, "w") as tempf:
            tempf.write("from test")
        result = ip.run_cell("%prql --file " + fname)
        assert result.result == [(1, "foo"), (2, "bar")]
        result = ip.run_cell("%prql -f " + fname)
        assert result.result == [(1, "foo"), (2, "bar")]


@pytest.mark.skip("We only support pandas")
def test_dict(ip):
    result = run_prql(ip, "from author")
    result = result.dict()
    assert isinstance(result, dict)
    assert "first_name" in result
    assert "last_name" in result
    assert "year_of_death" in result
    assert len(result["last_name"]) == 2


@pytest.mark.skip("We only support pandas")
def test_dicts(ip):
    result = run_prql(ip, "from author")
    for row in result.dicts():
        assert isinstance(row, dict)
        assert "first_name" in row
        assert "last_name" in row
        assert "year_of_death" in row


@pytest.mark.xfail(reason="Not supported in PRQL")
def test_bracket_var_substitution(ip):
    ip.user_global_ns["col"] = "first_name"
    assert run_prql(ip, "from author | filter WHERE {col} = 'William' ")[0] == (
        "William",
        "Shakespeare",
        1616,
    )

    ip.user_global_ns["col"] = "last_name"
    result = run_prql(ip, "from author | filter WHERE {col} = 'William' ")
    assert not result


# the next two tests had the same name, so I added a _2 to the second one
@pytest.mark.xfail(reason="Not supported in PRQL")
def test_multiline_bracket_var_substitution(ip):
    ip.user_global_ns["col"] = "first_name"
    assert run_prql(ip, "SELECT * FROM author\n" " WHERE {col} = 'William' ")[0] == (
        "William",
        "Shakespeare",
        1616,
    )

    ip.user_global_ns["col"] = "last_name"
    result = run_prql(ip, "SELECT * FROM author" " WHERE {col} = 'William' ")
    assert not result


@pytest.mark.xfail(reason="Not supported in PRQL")
def test_multiline_bracket_var_substitution_2(ip):
    ip.user_global_ns["col"] = "first_name"
    result = ip.run_cell_magic(
        "prql",
        "",
        """
        sqlite:// SELECT * FROM author
        WHERE {col} = 'William'
        """,
    )
    assert ("William", "Shakespeare", 1616) in result

    ip.user_global_ns["col"] = "last_name"
    result = ip.run_cell_magic(
        "prql",
        "",
        """
        sqlite:// SELECT * FROM author
        WHERE {col} = 'William'
        """,
    )
    assert not result


def test_json_in_select(ip):
    # Variable expansion does not work within json, but
    # at least the two usages of curly braces do not collide
    ip.user_global_ns["person"] = "prince"
    result = ip.run_cell_magic(
        "sql",
        "",
        """
        sqlite://
        SELECT
          '{"greeting": "Farewell sweet {person}"}'
        AS json
        """,
    )

    assert result == [('{"greeting": "Farewell sweet {person}"}',)]


def test_close_connection(ip):
    # TODO: change this to testing run_prql
    connections = run_sql(ip, "%sql -l")
    connection_name = list(connections)[0]
    run_sql(ip, f"%sql -x {connection_name}")
    connections_afterward = run_sql(ip, "%sql -l")
    assert connection_name not in connections_afterward


def test_pass_existing_engine(ip, tmp_empty, regtest):
    ip.user_global_ns["my_engine"] = create_engine("sqlite:///my.db")
    ip.run_line_magic("sql", "  my_engine ")

    run_sql(
        ip,
        [
            "CREATE TABLE some_data (n INT, name TEXT)",
            "INSERT INTO some_data VALUES (10, 'foo')",
            "INSERT INTO some_data VALUES (20, 'bar')",
        ],
    )

    result = ip.run_line_magic("prql", "from some_data")

    print(result, file=regtest)


# theres some weird shared state with this one, moving it to the end
def test_autolimit(ip):
    ip.run_line_magic("config", "SqlMagic.autolimit = 0")
    result = run_prql(ip, "from test")
    assert len(result) == 2
    ip.run_line_magic("config", "SqlMagic.autolimit = 1")
    result = run_prql(ip, "from test")
    assert len(result) == 1
    assert len(result) == 1
    assert len(result) == 1
