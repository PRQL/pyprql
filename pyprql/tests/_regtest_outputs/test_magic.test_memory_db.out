   n name
0  1  foo
1  2  bar
